File: python/src/venn_diagram_lab/share_distribution.py
"""Item Share Distribution: per-membership-count item totals.

Mirror of the webtool's ``src/utils/shareDistribution.ts``.
"""

from __future__ import annotations

import numpy as np

__all__ = ["item_share_distribution"]

_MATRIX_NDIM = 2


def item_share_distribution(matrix: np.ndarray) -> dict[int, int]:
    """Count items per set-membership level.

    Parameters
    ----------
    matrix
        Binary item-by-set matrix (rows = items, cols = sets, cells in {0, 1}).

    Returns
    -------
    dict
        ``{k: count}`` for ``k`` in ``1..n_sets``. All bins are present
        even when their count is zero. Rows that sum to zero are skipped
        (universe-rule violation; defensive).
    """
    if matrix.size == 0:
        n_sets = matrix.shape[1] if matrix.ndim == _MATRIX_NDIM else 0
        return {k: 0 for k in range(1, n_sets + 1)}

    n_sets = matrix.shape[1]
    row_sums = (matrix > 0).sum(axis=1).astype(int)
    counts: dict[int, int] = {k: 0 for k in range(1, n_sets + 1)}
    for k in row_sums:
        kk = int(k)
        if 1 <= kk <= n_sets:
            counts[kk] += 1
    return counts

File: packages/core/src/shareDistribution.ts
/**
 * Item Share Distribution: per-membership-count item totals.
 *
 * Given a binary item × set matrix (rows = items, columns = sets,
 * cells in {0, 1}), returns a Map keyed by k ∈ [1, nSets] mapping
 * to the number of items present in exactly k sets.
 *
 * Zero-valued bins are included for every k in [1, nSets].
 * Rows that sum to zero are ignored (universe-rule violation; defensive).
 */
export function itemShareDistribution(
  matrix: readonly (readonly number[])[],
  nSets: number,
): Map<number, number> {
  const counts = new Map<number, number>();
  for (let k = 1; k <= nSets; k++) counts.set(k, 0);

  for (const row of matrix) {
    let k = 0;
    for (let i = 0; i < nSets; i++) if (row[i]) k++;
    if (k >= 1 && k <= nSets) counts.set(k, (counts.get(k) ?? 0) + 1);
  }

  return counts;
}
